File: tools/component-tool/src/component_tool/scanner.py
"""Component file scanner - finds .component.md files under the workspace root."""

from __future__ import annotations

import subprocess
from pathlib import Path


def find_git_root(start: Path) -> Path | None:
    """Find the git repository root starting from the given directory."""
    try:
        result = subprocess.run(
            ["git", "rev-parse", "--show-toplevel"],
            cwd=str(start),
            capture_output=True,
            text=True,
            check=True,
        )
        return Path(result.stdout.strip())
    except (subprocess.CalledProcessError, FileNotFoundError):
        return None


def is_gitignored(path: Path, git_root: Path) -> bool:
    """Check if a file is gitignored."""
    try:
        result = subprocess.run(
            ["git", "check-ignore", "-q", str(path)],
            cwd=str(git_root),
            capture_output=True,
        )
        return result.returncode == 0
    except (subprocess.CalledProcessError, FileNotFoundError):
        return False


def scan_component_files(
    components_root: Path,
    git_root: Path | None = None,
) -> list[Path]:
    """Scan for .component.md files under components_root, excluding gitignored files.

    Returns sorted list of absolute paths.
    """
    if git_root is None:
        git_root = find_git_root(components_root)

    component_files: list[Path] = []
    for path in sorted(components_root.rglob("*.component.md")):
        if not path.is_file():
            continue
        # Resolve and check for symlink escapes
        resolved = path.resolve()
        try:
            resolved.relative_to(components_root.resolve())
        except ValueError:
            continue

        if git_root and is_gitignored(path, git_root):
            continue

        component_files.append(path)

    return component_files


def resolve_component_path(
    link_path: str,
    from_file: Path,
    components_root: Path,
) -> Path | None:
    """Resolve a relative link from a component file to an absolute path.

    Returns None if the resolved path is outside components_root or doesn't exist.
    """
    base_dir = from_file.parent
    resolved = (base_dir / link_path).resolve()

    # Check it's within components_root
    try:
        resolved.relative_to(components_root.resolve())
    except ValueError:
        return None

    if not resolved.is_file():
        return None

    return resolved

File: tools/component-tool-ts/src/scanner.ts
/**
 * Component file scanner - finds .component.md files under the workspace root.
 */

import { execSync } from "child_process";
import { existsSync, lstatSync, realpathSync } from "fs";
import { resolve, relative } from "path";
import { glob } from "./utils.js";

export function findGitRoot(start: string): string | null {
  try {
    const result = execSync("git rev-parse --show-toplevel", {
      cwd: start,
      encoding: "utf-8",
      stdio: ["pipe", "pipe", "pipe"],
    });
    return result.trim();
  } catch {
    return null;
  }
}

export function isGitignored(filePath: string, gitRoot: string): boolean {
  try {
    execSync(`git check-ignore -q ${JSON.stringify(filePath)}`, {
      cwd: gitRoot,
      stdio: ["pipe", "pipe", "pipe"],
    });
    return true;
  } catch {
    return false;
  }
}

export function scanComponentFiles(
  componentsRoot: string,
  gitRoot: string | null = null,
): string[] {
  if (gitRoot === null) {
    gitRoot = findGitRoot(componentsRoot);
  }

  const resolvedRoot = realpathSync(componentsRoot);
  const pattern = "**/*.component.md";
  const matches = glob(pattern, componentsRoot);

  const result: string[] = [];
  for (const match of matches) {
    const absPath = resolve(componentsRoot, match);

    // Check it's a file
    if (!existsSync(absPath) || !lstatSync(absPath).isFile()) continue;

    // Check for symlink escapes
    try {
      const realPath = realpathSync(absPath);
      const rel = relative(resolvedRoot, realPath);
      if (rel.startsWith("..")) continue;
    } catch {
      continue;
    }

    // Check gitignored
    if (gitRoot && isGitignored(absPath, gitRoot)) continue;

    result.push(absPath);
  }

  return result.sort();
}

export function resolveComponentPath(
  linkPath: string,
  fromFile: string,
  componentsRoot: string,
): string | null {
  const baseDir = resolve(fromFile, "..");
  const resolved = resolve(baseDir, linkPath);
  const resolvedRoot = realpathSync(componentsRoot);

  // Check within root
  const rel = relative(resolvedRoot, resolved);
  if (rel.startsWith("..") || resolve(resolvedRoot, rel) !== resolved) {
    return null;
  }

  if (!existsSync(resolved) || !lstatSync(resolved).isFile()) {
    return null;
  }

  return resolved;
}
